vertical-pod-autoscaler/enhancements/4902-delete-oom-pods/README.md

```diff
@@ -0,0 +1,79 @@
+# KEP-4902: Delete OOM Pods
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+   - [Goals](#goals)
+   - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+   - [Test Plan](#test-plan)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+   - [Update the eviction API](#update-the-eviction-api)
+<!-- /toc -->
+
+## Summary
+
+The default behaviour of VPA Updater is to evict Pods when new resource
+recommendations are available. This works fine for most cases.
+A problem that can arise is when there are multiple replicas and a
+PodDisruptionBudget (PDB) which allows one disruption. Now the eviction is at
+risk, because as soon as multiple replicas run into for example the memory
+limit and get killed they will never recover as the eviciton API will not allow
+any further disruptions.
+
+This proposal addresses the problem by allowing users to enable the deletion of
+pods as a backup if the eviction fails.
+
+## Motivation
+
+The motivation behind the change is to give VPA users a way to recover a
+failing deployment, if updated/increased limits would solve the problem.
+
+### Goals
+
+- Main: allow cluster administrators to enable deletion of pods
+- Secondary: configurable deletion threshold to tune the deletion behaviour
+
+### Non-Goals
+
+- Get rid of or work around the existing eviction behaviour
+
+## Proposal
+
+The proposal is to add `--delete-on-eviction-error` to the VPA to enable
```

Comment:
Why flags not API configuration? To make experiment easier to implement? Do you think there is little benefit to allowing to configure this on VPA-object level instead on a cluster level? Something else?

Reply:
> Do you think there is little benefit to allowing to configure this on VPA-object level instead on a cluster level

For us it would also be the case that we'd enable this for specific clusters entirely, not for single VPA objects. Even in the case we would make this configurable per VPA, having a global option (that people could override per VPA) would be preferred – having to touch each and every VPA is kind of cumbersome.

Reply:
> On the other hand there are people who don't manage their clusters (for example on GKE but also when many teams share a cluster) and then they can't change value of the flag.
> 
> While I understand it might be less convenient in some cases I think it's more important to make it possible at all.

That makes perfect sense. Are these two options mutually exclusive or would we want to have a global switch _and_ the option to configure this per VPA? There are already multiple cases where this pattern exists, right?

Reply:
> That makes perfect sense. Are these two options mutually exclusive or would we want to have a global switch _and_ the option to configure this per VPA? There are already multiple cases where this pattern exists, right?

have you seen the update I just pushed yesterday? I also added a field for the VPA resource in addition to the flag to the proposal. It's very similar to the minReplicas setting. I think this will be the best solution everyone.

Reply:
I think it's better to just do the API field. With both API field and flag we get the following semantics:

 - `DeleteOomingOnEvictionError = true` - delete pod if eviction failed and it's OOMing,
 - `DeleteOomingOnEvictionError = false` - don't delete pod even if it's OOMing and eviction fails (current behavior),
 - `DeleteOomingOnEvictionError` not set  - talk to your administrator or run a test to see what happens. But it can change.

I think having clean semantics on what happens by default is better than making (one time!) migration easier.

Reply:
If it would only be a one time migration it would be ok, but if we think of other open-source projects who have to potentially adjust their deployment mechanism to support such a setting it might take a very long time to get this done with. So I would still vote for having both.
Also asking your cluster administrator or looking into some documentation for your clusters doesn't sound that bad to me.